delta/plugins/archive/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/archive/ArchiveDownload.scala

```diff
@@ -0,0 +1,189 @@
+import ch.epfl.bluebrain.nexus.delta.sdk.model.acls.AclAddress
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{BaseUri, ResourceRef}
+import ch.epfl.bluebrain.nexus.delta.sdk.{Acls, AkkaSource, ReferenceExchange}
+import com.typesafe.scalalogging.Logger
+import monix.bio.{IO, UIO}
+
+/**
+  * Archive download functionality.
+  */
+trait ArchiveDownload {
+
+  /**
+    * Generates an akka [[Source]] of bytes representing the tar.
+    *
+    * @param value          the archive value
+    * @param project        the archive parent project
+    * @param ignoreNotFound do not fail when resource references are not found
+    * @param caller         the caller to be used for checking for access
+    */
+  def apply(
+      value: ArchiveValue,
+      project: ProjectRef,
+      ignoreNotFound: Boolean
+  )(implicit caller: Caller): IO[ArchiveRejection, AkkaSource]
+
+}
+
+object ArchiveDownload {
+
+  /**
+    * The default [[ArchiveDownload]] implementation.
+    *
+    * @param exchanges the collection of [[ReferenceExchange]] implementations
+    * @param acls      the acls module
+    * @param files     the files module
+    * @param sort      the configuration for sorting json keys
+    */
+  class ArchiveDownloadImpl(exchanges: Set[ReferenceExchange], acls: Acls, files: Files)(implicit
+      sort: JsonKeyOrdering,
+      baseUri: BaseUri,
+      rcr: RemoteContextResolution
+  ) extends ArchiveDownload {
+
+    implicit private val logger: Logger = Logger[ArchiveDownload]
+
+    override def apply(
+        value: ArchiveValue,
+        project: ProjectRef,
+        ignoreNotFound: Boolean
+    )(implicit caller: Caller): IO[ArchiveRejection, AkkaSource] = {
+      val referenceList = value.resources.value.toList
+      for {
+        _          <- checkResourcePermissions(referenceList, project)
+        optEntries <- referenceList.traverse {
+                        case ref: ResourceReference => fetchResource(ref, project, ignoreNotFound)
+                        case ref: FileReference     => fetchFile(ref, project, ignoreNotFound)
+                      }
+        entries     = optEntries.collect { case Some(value) => value } // discard None values
+        sorted      = entries.sortBy(_._1.filePath)
+      } yield Source(sorted).via(Archive.tar())
+    }
+
+    private def checkResourcePermissions(
+        refs: List[ArchiveReference],
+        project: ProjectRef
+    )(implicit caller: Caller): IO[AuthorizationFailed, Unit] = {
+      val authTargets = refs.collect { case ref: ResourceReference => ref }.map { ref =>
+        val p       = ref.project.getOrElse(project)
+        val address = AclAddress.Project(p)
+        (address, resources.read)
+      }
+      acls.authorizeForAny(authTargets).flatMap { authResult =>
+        authResult.find {
```

Comment:
This can be simplified to:

```scala
IO.fromEither(authResult.collectFirst{ case (addr, false) => addr}.toLeft(AuthorizationFailed(_, resources.read)))
```